pkg/storage/etcd/etcd_helper.go

```diff
@@ -40,15 +40,29 @@ import (
 	"golang.org/x/net/context"
 )
 
+// ValueTransformer allows a string value to be transformed before being read from or written to the underlying store.
+type ValueTransformer interface {
+	ReadString(string) (string, error)
+	WriteString(string) (string, error)
+}
+
+type identityTransformer struct{}
+
+func (identityTransformer) ReadString(s string) (string, error)  { return s, nil }
+func (identityTransformer) WriteString(s string) (string, error) { return s, nil }
+
+var IdentityTransformer ValueTransformer = identityTransformer{}
+
 // Creates a new storage interface from the client
 // TODO: deprecate in favor of storage.Config abstraction over time
-func NewEtcdStorage(client etcd.Client, codec runtime.Codec, prefix string, quorum bool, cacheSize int) storage.Interface {
+func NewEtcdStorage(client etcd.Client, codec runtime.Codec, prefix string, quorum bool, cacheSize int, transformer ValueTransformer) storage.Interface {
```

Comment:
Not sure what you mean by key filtering?  This was not an ACL thing, this was intended to enable encrypt at rest for resources as an optional thing.  Definitely comes with a cost, needed to get to a measurable point.

From a use case point of view the simplest and least impactful is simple encrypt at rest, while others (encrypt before kube, external secret integration) are all reasonable and can be done out of tree, but are more opinionated.  Someone willing to pay the operational cost for encrypt at rest can do so in tree.